File: apps/dashboard/app/api/chat/tools/audiences.ts
import Anthropic from '@anthropic-ai/sdk';
import { prisma } from '@workspace/database/client';
import type { ToolContext } from './ecommerce';

// ---------------------------------------------------------------------------
// Audience Tool Schemas
// ---------------------------------------------------------------------------

export const AUDIENCE_TOOL_SCHEMAS: Anthropic.Tool[] = [
  {
    name: 'list_audiences',
    description:
      'List all custom audiences, lookalike audiences, and saved audiences. Shows name, type, estimated size, platforms, and sync status.',
    input_schema: {
      type: 'object' as const,
      properties: {},
      required: [],
    },
  },
  {
    name: 'create_custom_audience',
    description:
      'Create a custom audience segment. Types: customer_list (hashed emails from orders), website (retargeting), catalog (product interaction). After creating, suggest creating a lookalike from it.',
    input_schema: {
      type: 'object' as const,
      properties: {
        audience_type: {
          type: 'string',
          enum: ['customer_list', 'website', 'catalog'],
          description: 'customer_list, website, or catalog',
        },
        name: {
          type: 'string',
          description: 'Audience name, e.g. "VIP Customers - Top 10%"',
        },
        description: {
          type: 'string',
          description: 'Description of the audience',
        },
        segment: {
          type: 'string',
          description: 'For customer_list: all, vip, repeat, high_ltv, recent, lapsed',
          default: 'all',
        },
        event_type: {
          type: 'string',
          description: 'For website/catalog: event to target (all_visitors, add_to_cart, purchase, etc.)',
          default: 'all_visitors',
        },
        retention_days: {
          type: 'integer',
          description: 'For website/catalog: days to look back (7, 14, 30, 60, 90, 180)',
          default: 30,
        },
        platforms: {
          type: 'array',
          items: { type: 'string' },
          description: 'Platforms to sync to: meta, google, bing',
        },
      },
      required: ['audience_type', 'name'],
    },
  },
  {
    name: 'create_lookalike_audience',
    description:
      'Create a lookalike audience from an existing custom audience. Meta/Google finds people similar to your source audience — best prospecting strategy.',
    input_schema: {
      type: 'object' as const,
      properties: {
        source_audience_id: {
          type: 'string',
          description: 'ID of the source audience to base the lookalike on',
        },
        name: {
          type: 'string',
          description: 'Name for the lookalike, e.g. "LAL 1% - VIP Customers - US"',
        },
        country: {
          type: 'string',
          description: 'ISO country code, e.g. US, CA, GB',
          default: 'US',
        },
        ratio: {
          type: 'number',
          description: 'Lookalike size 0.01-0.10 (1% = highest quality, 10% = broadest)',
          default: 0.01,
        },
      },
      required: ['source_audience_id', 'name'],
    },
  },
  {
    name: 'get_audience_insights',
    description:
      'Get estimated size, overlap data, and details for a specific audience segment.',
    input_schema: {
      type: 'object' as const,
      properties: {
        audience_id: {
          type: 'string',
          description: 'Audience segment ID',
        },
      },
      required: ['audience_id'],
    },
  },
  {
    name: 'smart_targeting',
    description:
      'Analyse customer order data to generate smart targeting recommendations: top geographies, AOV segments, customer lifetime value distribution. Use BEFORE creating a campaign to make data-driven targeting decisions.',
    input_schema: {
      type: 'object' as const,
      properties: {
        days: {
          type: 'integer',
          description: 'Days of order history to analyse (default 90)',
          default: 90,
        },
      },
      required: [],
    },
  },
  {
    name: 'search_locations',
    description:
      'Search for Meta geo targeting keys by location name. Use BEFORE creating a Meta campaign to resolve region/city names to the correct targeting keys.',
    input_schema: {
      type: 'object' as const,
      properties: {
        query: {
          type: 'string',
          description: 'Location name to look up, e.g. "California", "Ontario", "Toronto"',
        },
        location_type: {
          type: 'string',
          enum: ['region', 'city', 'country', 'zip'],
          description: 'region, city, country, or zip',
          default: 'region',
        },
      },
      required: ['query'],
    },
  },
];

export const AUDIENCE_TOOL_NAMES = new Set(AUDIENCE_TOOL_SCHEMAS.map((t) => t.name));

// ---------------------------------------------------------------------------
// Static location key lookup (subset — full lookup would call Meta API)
// ---------------------------------------------------------------------------

const REGION_KEYS: Record<string, { key: string; name: string; country: string }> = {
  california: { key: '3847', name: 'California', country: 'US' },
  texas: { key: '3861', name: 'Texas', country: 'US' },
  florida: { key: '3851', name: 'Florida', country: 'US' },
  'new york': { key: '3855', name: 'New York', country: 'US' },
  illinois: { key: '3852', name: 'Illinois', country: 'US' },
  washington: { key: '3863', name: 'Washington', country: 'US' },
  ontario: { key: '533', name: 'Ontario', country: 'CA' },
  'british columbia': { key: '536', name: 'British Columbia', country: 'CA' },
  alberta: { key: '534', name: 'Alberta', country: 'CA' },
  'england': { key: '2347', name: 'England', country: 'GB' },
  'scotland': { key: '2348', name: 'Scotland', country: 'GB' },
  'maharashtra': { key: '2348', name: 'Maharashtra', country: 'IN' },
  'delhi': { key: '2349', name: 'Delhi', country: 'IN' },
  'karnataka': { key: '2350', name: 'Karnataka', country: 'IN' },
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function estimateAudienceSize(
  audienceType: string,
  segment: string,
  eventType: string,
  retentionDays: number,
  totalCustomers: number
): number {
  let base = totalCustomers;
  if (audienceType === 'customer_list') {
    switch (segment) {
      case 'vip':     base = Math.round(totalCustomers * 0.1); break;
      case 'repeat':  base = Math.round(totalCustomers * 0.3); break;
      case 'high_ltv': base = Math.round(totalCustomers * 0.2); break;
      case 'recent':  base = Math.round(totalCustomers * 0.25); break;
      case 'lapsed':  base = Math.round(totalCustomers * 0.4); break;
      default:        base = totalCustomers;
    }
    // Meta matches ~60% of emails
    return Math.round(base * 0.6);
  }

  if (audienceType === 'website') {
    const dayMultiplier = retentionDays / 30;
    const eventMultiplier = eventType === 'all_visitors' ? 10 : eventType === 'add_to_cart' ? 3 : 1;
    return Math.round(totalCustomers * eventMultiplier * dayMultiplier * 2);
  }

  if (audienceType === 'catalog') {
    return Math.round(totalCustomers * 2);
  }

  return totalCustomers;
}

// ---------------------------------------------------------------------------
// Tool handlers
// ---------------------------------------------------------------------------

async function handleListAudiences(ctx: ToolContext): Promise<unknown> {
  const segments = await prisma.audienceSegment.findMany({
    where: { organizationId: ctx.orgId },
    orderBy: { createdAt: 'desc' },
    take: 30,
    select: {
      id: true,
      name: true,
      description: true,
      type: true,
      platforms: true,
      estimatedSize: true,
      syncStatus: true,
      lastSyncAt: true,
      createdAt: true,
    },
  });

  return {
    total: segments.length,
    audiences: segments.map((s) => ({
      id: s.id,
      name: s.name,
      description: s.description,
      type: s.type,
      platforms: s.platforms,
      estimated_size: s.estimatedSize,
      sync_status: s.syncStatus,
      last_sync_at: s.lastSyncAt?.toISOString().slice(0, 10) ?? null,
      created_at: s.createdAt.toISOString().slice(0, 10),
    })),
  };
}

async function handleCreateCustomAudience(
  input: Record<string, unknown>,
  ctx: ToolContext
): Promise<unknown> {
  const {
    audience_type,
    name,
    description,
    segment = 'all',
    event_type = 'all_visitors',
    retention_days = 30,
    platforms = ['meta'],
  } = input as {
    audience_type: string;
    name: string;
    description?: string;
    segment?: string;
    event_type?: string;
    retention_days?: number;
    platforms?: string[];
  };

  const totalCustomers = await prisma.customerProfile.count({
    where: { organizationId: ctx.orgId },
  });

  const estimatedSize = estimateAudienceSize(
    audience_type,
    segment,
    event_type,
    retention_days,
    totalCustomers
  );

  const rules =
    audience_type === 'customer_list'
      ? [{ field: 'segment', operator: 'equals', value: segment }]
      : audience_type === 'website'
      ? [
          { field: 'event_type', operator: 'equals', value: event_type },
          { field: 'retention_days', operator: 'equals', value: retention_days },
        ]
      : [{ field: 'catalog_event', operator: 'equals', value: event_type }];

  const newSegment = await prisma.audienceSegment.create({
    data: {
      organizationId: ctx.orgId,
      name,
      description: description ?? '',
      type: audience_type,
      platforms: platforms as string[],
      rules,
      estimatedSize,
      syncStatus: 'pending',
    },
    select: { id: true, name: true, estimatedSize: true },
  });

  return {
    success: true,
    audience_id: newSegment.id,
    name: newSegment.name,
    estimated_size: newSegment.estimatedSize,
    type: audience_type,
    platforms,
    sync_status: 'pending',
    message: `Custom audience "${name}" created with ~${estimatedSize.toLocaleString()} estimated members. Sync to platforms is queued.`,
    next_steps: 'Consider creating a 1% lookalike from this audience for prospecting using create_lookalike_audience.',
  };
}

async function handleCreateLookalikeAudience(
  input: Record<string, unknown>,
  ctx: ToolContext
): Promise<unknown> {
  const {
    source_audience_id,
    name,
    country = 'US',
    ratio = 0.01,
  } = input as {
    source_audience_id: string;
    name: string;
    country?: string;
    ratio?: number;
  };

  const source = await prisma.audienceSegment.findFirst({
    where: { id: source_audience_id, organizationId: ctx.orgId },
    select: { name: true, estimatedSize: true, platforms: true },
  });

  if (!source) return { error: 'Source audience not found' };

  const estimatedSize = Math.round((source.estimatedSize ?? 10000) / ratio * 0.01);

  const lookalike = await prisma.audienceSegment.create({
    data: {
      organizationId: ctx.orgId,
      name,
      description: `Lookalike of "${source.name}" — ${(ratio * 100).toFixed(0)}% in ${country}`,
      type: 'lookalike',
      platforms: source.platforms,
      rules: [
        { field: 'source_audience_id', operator: 'equals', value: source_audience_id },
        { field: 'country', operator: 'equals', value: country },
        { field: 'ratio', operator: 'equals', value: ratio },
      ],
      estimatedSize,
      syncStatus: 'pending',
    },
    select: { id: true, name: true, estimatedSize: true },
  });

  return {
    success: true,
    audience_id: lookalike.id,
    name: lookalike.name,
    source_audience: source.name,
    country,
    ratio_pct: (ratio * 100).toFixed(0) + '%',
    estimated_size: estimatedSize.toLocaleString(),
    message: `Lookalike audience "${name}" created. Meta will find ${(ratio * 100).toFixed(0)}% of ${country} Facebook users most similar to "${source.name}".`,
  };
}

async function handleGetAudienceInsights(
  input: Record<string, unknown>,
  ctx: ToolContext
): Promise<unknown> {
  const { audience_id } = input as { audience_id: string };

  const segment = await prisma.audienceSegment.findFirst({
    where: { id: audience_id, organizationId: ctx.orgId },
    include: {
      members: { select: { profileId: true }, take: 1 },
      _count: { select: { members: true } },
    },
  });

  if (!segment) return { error: 'Audience not found' };

  return {
    id: segment.id,
    name: segment.name,
    description: segment.description,
    type: segment.type,
    platforms: segment.platforms,
    estimated_size: segment.estimatedSize,
    actual_member_count: segment._count.members,
    sync_status: segment.syncStatus,
    synced_platforms: segment.syncedPlatforms,
    last_sync_at: segment.lastSyncAt?.toISOString() ?? null,
    rules: segment.rules,
    rule_logic: segment.ruleLogic,
  };
}

async function handleSmartTargeting(
  input: Record<string, unknown>,
  ctx: ToolContext
): Promise<unknown> {
  const days = (input.days as number) ?? 90;
  const since = new Date();
  since.setDate(since.getDate() - days);

  const orders = await prisma.commerceOrder.findMany({
    where: { organizationId: ctx.orgId, placedAt: { gte: since } },
    select: { totalAmount: true, channel: true },
    take: 1000,
  });

  if (orders.length === 0) {
    return {
      message: 'No order data found. Connect a commerce store to get smart targeting recommendations.',
      recommendations: [],
    };
  }

  // Group by channel; geography would require enriched address data
  const geoMap: Record<string, number> = {};
  const aovBuckets = { low: 0, mid: 0, high: 0 };
  let totalRevenue = 0;

  for (const o of orders) {
    const geo = o.channel ?? 'Direct';
    geoMap[geo] = (geoMap[geo] ?? 0) + 1;

    const price = parseFloat(o.totalAmount.toString());
    totalRevenue += price;
    if (price < 50) aovBuckets.low++;
    else if (price < 150) aovBuckets.mid++;
    else aovBuckets.high++;
  }

  const topGeos = Object.entries(geoMap)
    .sort(([, a], [, b]) => b - a)
    .slice(0, 5)
    .map(([name, count]) => ({ name, order_count: count, share_pct: ((count / orders.length) * 100).toFixed(1) }));

  const avgOrderValue = totalRevenue / orders.length;
  const dominantAovSegment =
    aovBuckets.high > aovBuckets.mid && aovBuckets.high > aovBuckets.low
      ? 'high (>$150)'
      : aovBuckets.mid > aovBuckets.low
      ? 'mid ($50-$150)'
      : 'low (<$50)';

  const recommendations = [
    `Target top geographic regions: ${topGeos.map((g) => g.name).join(', ')}`,
    `Suggested daily budget: $${Math.round(avgOrderValue * 0.2 * 10) / 10}-$${Math.round(avgOrderValue * 0.5 * 10) / 10} (based on ${ctx.currency} ${avgOrderValue.toFixed(0)} AOV)`,
    `Customer AOV profile: ${dominantAovSegment} — adjust bidding strategy accordingly`,
    `Create a customer list from your top buyers and use 1% lookalike for prospecting`,
  ];

  return {
    period: `${days}d`,
    total_orders: orders.length,
    avg_order_value: avgOrderValue.toFixed(2),
    currency: ctx.currency,
    top_regions: topGeos,
    aov_distribution: aovBuckets,
    recommendations,
  };
}

async function handleSearchLocations(input: Record<string, unknown>): Promise<unknown> {
  const query = ((input.query as string) ?? '').toLowerCase().trim();
  const locationType = (input.location_type as string) ?? 'region';

  const match = REGION_KEYS[query];
  if (match) {
    return {
      results: [{ ...match, type: locationType }],
      note: 'Use the "key" value in your campaign targeting spec.',
    };
  }

  // Fuzzy partial match
  const fuzzy = Object.values(REGION_KEYS).filter((v) =>
    v.name.toLowerCase().includes(query)
  );

  if (fuzzy.length > 0) {
    return {
      results: fuzzy.map((v) => ({ ...v, type: locationType })),
      note: 'Use the "key" value in your campaign targeting spec.',
    };
  }

  return {
    results: [],
    message: `No targeting key found for "${query}". For production, this would call the Meta Targeting Search API. For now, use the country ISO code directly (e.g. US, CA, GB) as a fallback.`,
  };
}

// ---------------------------------------------------------------------------
// Dispatcher
// ---------------------------------------------------------------------------

export async function executeAudienceTool(
  name: string,
  input: Record<string, unknown>,
  ctx: ToolContext
): Promise<unknown> {
  switch (name) {
    case 'list_audiences':           return handleListAudiences(ctx);
    case 'create_custom_audience':   return handleCreateCustomAudience(input, ctx);
    case 'create_lookalike_audience': return handleCreateLookalikeAudience(input, ctx);
    case 'get_audience_insights':    return handleGetAudienceInsights(input, ctx);
    case 'smart_targeting':          return handleSmartTargeting(input, ctx);
    case 'search_locations':         return handleSearchLocations(input);
    default:
      throw new Error(`Unknown audience tool: ${name}`);
  }
}

File: services/accelerate-agentic-service/src/agentic_platform/app/accelera/tools/audiences.py
"""Audience tools — 6 tools that call accelerate-cdp-service.

Audience size estimation logic ported from Next.js /api/chat/tools/audiences.ts.
"""

from langchain_core.tools import tool

from src.agentic_platform.core.engine import AgenticTool, ToolResponse, ToolTag
from src.agentic_platform.core.engine.context import get_org_id
from src.agentic_platform.app.accelera.services.clients import cdp_client
from src.agentic_platform.app.accelera.models.audiences import REGION_KEYS
from src.agentic_platform.app.accelera.blocks import (
    audience_card_block, AudienceCardData, AudienceRow,
    metric_cards_block, MetricCardsData, MetricItem,
)


def _estimate_audience_size(audience_type: str, event_type: str = "", source_size: int = 0, ratio: float = 0.01) -> int:
    """Estimate audience size (ported from Next.js estimation logic)."""
    match audience_type:
        case "customer_list": return int(source_size * 0.60)      # 60% email match
        case "website":
            match event_type:
                case "all_visitors": return source_size * 10
                case "add_to_cart": return source_size * 3
                case _: return source_size * 5
        case "catalog": return source_size * 2
        case "lookalike": return int(source_size / ratio * 0.01)
        case _: return source_size


# ── list_audiences ────────────────────────────────────────────────────

@tool("list_audiences")
async def _list_audiences() -> dict:
    """List all custom, lookalike, and saved audience segments.
    Use when the user asks about their audiences or wants to see available targeting options."""
    org_id = get_org_id()
    resp = await cdp_client.get(f"/segments?org_id={org_id}")
    body = resp.get("body", {})
    segments = body.get("segments", body) if isinstance(body, dict) else body

    rows = [
        AudienceRow(
            id=s.get("id", ""),
            name=s.get("name", ""),
            type=s.get("type", "custom"),
            estimated_size=s.get("estimatedSize") or s.get("estimated_size", 0),
            status=s.get("status", "ready"),
            platform=s.get("platform", "meta"),
        )
        for s in (segments if isinstance(segments, list) else [])
    ]

    return ToolResponse(
        summary=f"Found {len(rows)} audience segments.",
        data={"audiences": [r.model_dump() for r in rows]},
        ui_blocks=[audience_card_block.create(data=AudienceCardData(audiences=rows, total_count=len(rows)))],
    ).model_dump()

list_audiences = AgenticTool(func=_list_audiences, thinking_messages=["Loading your audience segments..."], tags=[ToolTag.ANALYTICS])


# ── create_custom_audience ────────────────────────────────────────────

@tool("create_custom_audience")
async def _create_custom_audience(
    name: str,
    audience_type: str,
    platform: str = "meta",
    event_type: str = "all_visitors",
    source_size: int = 1000,
) -> dict:
    """Create a custom audience segment (customer_list, website, or catalog).
    Use when the user asks to create an audience or set up retargeting."""
    org_id = get_org_id()
    estimated_size = _estimate_audience_size(audience_type, event_type, source_size)

    resp = await cdp_client.post("/segments", json={
        "org_id": org_id,
        "name": name,
        "type": audience_type,
        "platform": platform,
        "event_type": event_type,
        "estimated_size": estimated_size,
    })
    body = resp.get("body", {})

    return ToolResponse(
        summary=f"Created '{name}' audience (~{estimated_size:,} people).",
        data={"audience": body, "estimated_size": estimated_size},
    ).model_dump()

create_custom_audience = AgenticTool(func=_create_custom_audience, thinking_messages=["Creating your audience..."], tags=[ToolTag.ANALYTICS])


# ── create_lookalike_audience ─────────────────────────────────────────

@tool("create_lookalike_audience")
async def _create_lookalike_audience(
    name: str,
    source_audience_id: str,
    platform: str = "meta",
    ratio: float = 0.01,
) -> dict:
    """Create a lookalike audience from an existing source audience.
    Use when the user wants to find new customers similar to existing ones."""
    org_id = get_org_id()
    resp = await cdp_client.post("/segments", json={
        "org_id": org_id,
        "name": name,
        "type": "lookalike",
        "platform": platform,
        "source_id": source_audience_id,
        "ratio": ratio,
    })
    body = resp.get("body", {})
    estimated_size = body.get("estimatedSize") or body.get("estimated_size", 0)

    return ToolResponse(
        summary=f"Created lookalike audience '{name}' (~{estimated_size:,} people, {ratio*100:.0f}% similarity).",
        data={"audience": body},
    ).model_dump()

create_lookalike_audience = AgenticTool(func=_create_lookalike_audience, thinking_messages=["Building lookalike audience..."], tags=[ToolTag.ANALYTICS])


# ── get_audience_insights ─────────────────────────────────────────────

@tool("get_audience_insights")
async def _get_audience_insights(audience_id: str) -> dict:
    """Get insights for a specific audience: size, overlap data, and sync status.
    Use when the user asks about a specific audience's details or performance."""
    resp = await cdp_client.get(f"/segments/{audience_id}")
    body = resp.get("body", {})

    size = body.get("estimatedSize") or body.get("estimated_size", 0)
    name = body.get("name", audience_id)
    status = body.get("status", "unknown")

    metrics = [
        MetricItem(label="Audience Size", value=f"{size:,}"),
        MetricItem(label="Status", value=status.capitalize()),
        MetricItem(label="Type", value=body.get("type", "custom").capitalize()),
        MetricItem(label="Platform", value=body.get("platform", "meta").capitalize()),
    ]

    return ToolResponse(
        summary=f"Audience '{name}': {size:,} people, status: {status}.",
        data=body,
        ui_blocks=[metric_cards_block.create(data=MetricCardsData(metrics=metrics, title=f"Audience: {name}"))],
    ).model_dump()

get_audience_insights = AgenticTool(func=_get_audience_insights, thinking_messages=["Fetching audience insights..."], tags=[ToolTag.ANALYTICS])


# ── smart_targeting ───────────────────────────────────────────────────

@tool("smart_targeting")
async def _smart_targeting() -> dict:
    """Generate data-driven targeting recommendations based on order history.
    Use when the user asks for targeting suggestions or audience strategy."""
    org_id = get_org_id()
    resp = await cdp_client.post(f"/segments/compute", json={"org_id": org_id, "type": "smart_targeting"})
    body = resp.get("body", {})

    geo = body.get("top_regions") or body.get("geo", [])
    aov_segment = body.get("aov_segment", "mid_tier")
    ltv = body.get("ltv_estimate", 0.0)

    metrics = [
        MetricItem(label="Top Regions", value=", ".join(geo[:3]) if geo else "N/A"),
        MetricItem(label="Customer Segment", value=aov_segment.replace("_", " ").title()),
        MetricItem(label="Est. LTV", value=f"${ltv:,.2f}"),
    ]

    return ToolResponse(
        summary=f"Smart targeting: focus on {', '.join(geo[:2]) or 'top regions'}, {aov_segment} AOV segment.",
        data=body,
        ui_blocks=[metric_cards_block.create(data=MetricCardsData(metrics=metrics, title="Smart Targeting Recommendations"))],
    ).model_dump()

smart_targeting = AgenticTool(func=_smart_targeting, thinking_messages=["Analyzing your customer data for targeting..."], tags=[ToolTag.ANALYTICS])


# ── search_locations ──────────────────────────────────────────────────

@tool("search_locations")
async def _search_locations(query: str) -> dict:
    """Resolve location names to Meta geo-targeting keys for audience and campaign setup.
    Use when the user asks about targeting specific locations or geographic regions."""
    query_lower = query.lower().strip()
    matches = []
    for key, loc in REGION_KEYS.items():
        if query_lower in key or key in query_lower:
            matches.append(loc.model_dump())

    if not matches:
        return ToolResponse(
            summary=f"No exact match for '{query}'. Try country or region names like 'United States', 'California', 'United Kingdom'.",
            data={"query": query, "matches": []},
        ).model_dump()

    return ToolResponse(
        summary=f"Found {len(matches)} location(s) matching '{query}'.",
        data={"query": query, "matches": matches},
    ).model_dump()

search_locations = AgenticTool(func=_search_locations, thinking_messages=["Looking up location targeting keys..."], tags=[ToolTag.ANALYTICS])
